crates/library/std/src/evm.fe

```diff
@@ -268,24 +270,24 @@ pub unsafe fn ext_code_hash(_ addr: address) -> u256 {
     return __extcodehash(u256(addr))
 }
 
-pub unsafe fn keccak256_mem(offset: u256, len: u256) -> u256 {
-    return __keccak256(offset, len)
+pub fn keccak256_mem(buffer: buf::MemoryBuffer) -> u256 {
+   unsafe { return __keccak256(buffer.offset(), buffer.len()) }
 }
 
 
 // Contract creation and calling
 
-pub unsafe fn create(value v: u256, offset p: u256, len n: u256) -> address {
-    return address(__create(v, p, n))
+pub fn create(value v: u256, buffer: buf::MemoryBuffer) -> address {
+    unsafe { return address(__create(v, buffer.offset(), buffer.len())) } 
 }
 
-pub unsafe fn create2(value v: u256, offset p: u256, len n: u256, salt s: u256) -> address {
-    return address(__create2(v, p, n, s))
+pub fn create2(value v: u256, buffer: buf::MemoryBuffer, salt s: u256) -> address {
+    unsafe { return address(__create2(v, buffer.offset(), buffer.len(), s)) }
 }
 
 // TODO: return bool (success)
-pub unsafe fn call(gas: u256, addr: address, value: u256, input_offset: u256, input_len: u256, output_offset: u256, output_len: u256) -> u256 {
-    return __call(gas, u256(addr), value, input_offset, input_len, output_offset, output_len)
+pub fn call(gas: u256, addr: address, value: u256, buffer: buf::RawCallBuffer) -> bool {
+  unsafe{ return __call(gas, u256(addr), value, buffer.offset(), buffer.input_len(), buffer.offset(), buffer.output_len()) == 1 }
```

Comment:
the `buffer` parameter doesnt actually need to be mutable here for this to work, but `__call` will in fact mutate the buffer, which is why the `buf` param on `raw_call` was `mut` and should be here as well

Reply:
since we are using `unsafe`, we need to make sure the semantics of the language still hold

Reply:
Is it enough to check changes in the compiled JSON files, or do we need to deploy to a testnet ?

Reply:
in reference to the snapshot changes?

Reply:
in reference to the semantic changes.

Reply:
ok, Im not quite sure I follow.. could you please explain further?

Reply:
So one of the comments entailed that we need to check the language semantics # # - because we are using unsafe, how would we go about testing/checking this ??

Reply:
oh, what I mean is that the implementation of `evm::call` will mutate the `buf` parameter, even though `buf` does not need to be mutable for it's use in the `__call` intrinsic

it's important that `mut` semantics hold when using `evm::call`, so we need to set the `buf` parameter as mutable, that way someone using the `evm::call` function will have to set the `buf` to mutable.

so..

```
pub fn call(gas: u256, addr: address, value: u256, buf: buf::RawCallBuffer) -> bool {
  unsafe{ return __call(gas, u256(addr), value, buffer.offset(), buf.input_len(), buf.offset(), buffer.output_len()) == 1 }
```

becomes

```
pub fn call(gas: u256, addr: address, value: u256, mut buf: buf::RawCallBuffer) -> bool {
  unsafe{ return __call(gas, u256(addr), value, buffer.offset(), buf.input_len(), buf.offset(), buffer.output_len()) == 1 }
}
```

that way someone writing the following

```
let buf: RawCallBuffer = ...
evm::call(..., buf)
```

will get a compile error asking them to make `buf` mutable, like so

```
let mut buf: RawCallBuffer = ...
evm::call(..., buf)
```

---

`unsafe` code allows you to do things that are normally disallowed in safe code, such as direct memory mutation, which is what we are doing in `evm::call`. this is fine as long as we correctly set the parameters that end up getting mutated as `mut`